### ISSUE TITLE
[Docs] Use full package names in `.hintrc` examples

### ISSUE THREAD

**Issue by antross** [webhintio/hint#2774]
Once #2640 is merged, we'll have support for specifying full package names in `.hintrc` files. This was necessary to support scoped packages names like `@foo/webhint-hint-bar`, but also raises the question of whether we should follow this pattern for all resources in `.hintrc` files for consistency between 1st and 3rd party resources as well as with what's listed in `package.json` (originally suggested by @gesa). 

In theory this could help reduce confusion between package names and resource names. As @gesa has pointed out, the current mapping once #2640 merges ends up looking like:
>   * given `webhint-type-name`, the string is `name`
>   * `@hint/type-name` string is `name`
>   * `@foo/webhint-type-name` string is `@foo/webhint-type-name`

The first two options were present before and already a bit confusing as they require users to know which part of a package name is relevant in a `.hintrc` file. The third option, while necessary to enable scoped package names, adds to the confusion.

Example `.hintrc` today:
```jsonc
{
    "connector": "jsdom", // package @hint/connector-jsdom
    "extends": ["web-recommended"], // package @hint/configuration-web-recommended
    "hints": {
        "button-type": "error", // package @hint/hint-button-type
        "foo": "error", // package webhint-hint-foo
        "@foo/webhint-hint-bar": "error" // package @foo/webhint-hint-bar
    }
}
```

Example `.hintrc` with switch to full package names:
```json
{
    "connector": "@hint/connector-jsdom",
    "extends": ["@hint/configuration-web-recommended"],
    "hints": {
        "@hint/hint-button-type": "error",
        "webhint-hint-foo": "error",
        "@foo/webhint-hint-bar": "error"
    }
}
```

Note no further code change is required as #2640 enables support for both `.hintrc` examples above.

TL;DR As @gesa originally suggested, we can improve the user experience by updating our documentation and configurations to always list full package names for resources (while still supporting short names). Then we can eventually consider simplifying the resource loader by requiring full package names in the future.

@webhintio/contributors Any thoughts or objections?

**Comment by gesa** [comment]
On a related note, I have a PR on @antross's fork that I plan to close and re-open here when #2640 merges. I've done a bit of work on that aspect of the docs so this issue would require limited actual work from the core team. 😂  Mostly just reviewing.

**Comment by molant** [comment]
I'm a bit concerned about the verbosity of having only full package names tbh.
Do we have stats about custom package usage?
Maybe instead of switching our packages we could improve the documentation for resource loading, the error messages when a package is not found, and point to the right place to debug it?



________________________________

**Comment by antross** [comment]
> I'm a bit concerned about the verbosity of having only full package names tbh.

@hxlnt any thoughts on this? Perhaps this is a good customer development opportunity.

Personally I'm more concerned about the inconsistency versus what's specified in `package.json`. I think "shorter" isn't necessarily better here even when we're only using first party packages as it requires users to understand which part of a package name is important. But, I could be worrying about something that doesn't phase most users.

**Comment by gesa** [comment]
I agree with @antross regarding knowing which part of a package name is important, it took me a second to put that all together.

Additionally though, i thought part of webhint's value prop was its extensibility (and specifically i'm thinking of being able to extend someone else's config a la airbnb as so many use for eslint). As it stands right now, figuring out how to use packages not specifically mentioned on the webhint docs is tricky. regardless of current usage stats, i wonder if more custom packages would be created and consumed if it were a little clearer as to how to use such things?

**Comment by molant** [comment]
> i wonder if more custom packages would be created and consumed if it were a little clearer as to how to use such things?

I don't disagree about improving the docs and making things clearer. I'm just concerned about using only the full names for all the packages.

Maybe we should take a look at other projects to see if there's something else we should do (there's a new formar proposed for eslint IIRC), or if improving the docs is good enough for now.


________________________________

**Comment by hxlnt** [comment]
@molant Can you share a bit more about the concern around using full names? It's just a few characters per line, no? Just want to make sure I understand the concern. And agree that looking to other projects for precedent is a good call.

FWIW, I prefer the approach that follows the simplest-to-explain pattern that doesn't have a lot of "if this, then that" diversion. To me, that _seems_ like the full name approach. But maybe I'm missing something.

**Comment by molant** [comment]
It's more of following DRY.
Even if it is just a few characters per line it adds. In the very first versions we had "hint-" prepended. We didn't like the experience and changed it.

I think most of the people are using first party packages and we shouldn't make that experience worse.

I think we should improve the docs and explain better how people can create custom configurations and extend them.





________________________________

**Comment by antross** [comment]
@molant Is there a way to resolve name collisions in the short name case?

E.g. if I have a first party `@hint/hint-foo` and a third-party `webhint-hint-foo` can I:
1. Load them both
2. Pass options to both of them

**Comment by molant** [comment]
IIRC first party have higher priority when using short names.
If a developer decides to create a hint with the same name as a first party one and use both hints (or have both installed in the same project) then the full name should be used for disambiguation. Don't  know if we support it at the moment but if not we can update the docs to alert the user about it.

My main concern here is that I don't want to lose the option to use shortnames 😊

________________________________

**Comment by antross** [comment]
> My main concern here is that I don't want to lose the option to use shortnames

@molant to be clear I'm only proposing changing our examples, not our implementation, so shortnames would still be an option, just not the recommended default 😊

> then the full name should be used for disambiguation

Yes, now that we have support for full package names we could optionally use them to disambiguate as you suggest. However, there's also an issue with `hintOptions` disambiguation. The `hintOptions` are looked up using the hint ID from the hint metadata and searching for a partial match in the configuration from `.hintrc` 😱 

So we could still end up passing the same `hintOptions` to both hints. That is, unless the third party hint additionally changes its metadata to make its ID a full package name. But if it does that then the third party hint MUST be referenced using the full package name in `.hintrc` or we won't load it 😅

**Comment by antross** [comment]
@molant - I'd like to drive this to a resolution so we can update the docs. 

If you still object to using full package names everywhere in the documentation, then I propose we continue using short names for first-party resources for now and use full package names in all examples referencing third-party resources. Then we can revisit the discussion when you have more time.

**Comment by molant** [comment]
I still prefer to use shortnames for first party. Otherwise we will have to change also the output of create-hintrc to be consistent. Also we should document the edge cases of name collision, how to load a third party hint, etc.

Sounds good?


________________________________

**Comment by gesa** [comment]
So @antross, @molant, @hxlnt Is this considered agreed/concluded then?

These are my impressions, presume each to be a question, not a decree:

- Docs should still favor short names for first party (`@hint/`-scoped packages)
  - This will require a brief explanation in the concepts section explaining how to deduce the short name based on the package name
- Docs should explicitly explain how to load third-party hints
- Docs should explicitly explain what happens in case of name collision
  - Some level of detail needs to go into end user docs
  - Is this mostly a jsdoc task?
- Scoped packages outside of `@hint/` are not yet actually supported
  - There also doesn't _appear_ to be an item in the issue queue about this

**Comment by antross** [comment]
I'll let @molant confirm, but I think we've agreed to favor short names for first-party packages and full names for third-party packages. Sorry for not making this clear earlier.

With that in mind:
> Docs should still favor short names for first party (@hint/-scoped packages)
> * This will require a brief explanation in the concepts section explaining how to deduce the short name based on the package name

Yes and yes.

> Docs should explicitly explain how to load third-party hints

Yes.

> Docs should explicitly explain what happens in case of name collision

I think we can skip this one. Name collisions shouldn't happen with this approach so long as we also tell authors of third-party hints to use the full package name as the `id` in their hint metadata.

> Scoped packages outside of @hint/ are not yet actually supported

We shipped support a few weeks ago in `hint 5.2.0` (via f36055c685d7f7361dc84556e44a78ccbca246be). Let me know if this isn't working for you, though.

**Comment by molant** [comment]
>  I think we've agreed to favor short names for first-party packages and full names for third-party packages

Yup. Thanks!


________________________________
